tests/test_upgrades.cairo

```diff
@@ -0,0 +1,193 @@
+    return ();
+}
+
+@external
+func test_upgrade_implementation_non_admin{syscall_ptr: felt*, range_check_ptr}() {
+    tempvar factory_address;
+    tempvar router_address;
+    tempvar pair_address;
+    tempvar declared_factory_v2_class_hash;
+    tempvar declared_router_v2_class_hash;
+    tempvar declared_pair_v2_class_hash;
+
+    %{ 
+        ids.factory_address = context.factory_address
+        ids.router_address = context.router_address
+        ids.pair_address = context.pair_address
+        ids.declared_factory_v2_class_hash = declare("contracts/test/FactoryV2.cairo").class_hash
+        ids.declared_router_v2_class_hash = declare("contracts/test/RouterV2.cairo").class_hash
+        ids.declared_pair_v2_class_hash = declare("contracts/test/PairV2.cairo").class_hash
+    %}
+
+    %{ expect_revert(error_message="Proxy: caller is not admin") %}
+    IProxy.upgrade(contract_address=factory_address, new_implementation=declared_factory_v2_class_hash);
+
+    %{ expect_revert(error_message="Proxy: caller is not admin") %}
+    IProxy.upgrade(contract_address=router_address, new_implementation=declared_router_v2_class_hash);
+
+    %{ expect_revert(error_message="Proxy: caller is not admin") %}
+    IProxy.upgrade(contract_address=pair_address, new_implementation=declared_pair_v2_class_hash);
+
+    return ();
+}
+
+@external
+func test_upgrade_implementation{syscall_ptr: felt*, range_check_ptr}() {
+    tempvar deployer_address;
+    tempvar factory_address;
+    tempvar router_address;
+    tempvar pair_address;
+    tempvar declared_factory_v2_class_hash;
+    tempvar declared_router_v2_class_hash;
+    tempvar declared_pair_v2_class_hash;
+
+    %{
+        ids.deployer_address = context.deployer_address
+        ids.factory_address = context.factory_address
+        ids.router_address = context.router_address
+        ids.pair_address = context.pair_address
+        ids.declared_factory_v2_class_hash = declare("contracts/test/FactoryV2.cairo").class_hash
+        ids.declared_router_v2_class_hash = declare("contracts/test/RouterV2.cairo").class_hash
+        ids.declared_pair_v2_class_hash = declare("contracts/test/PairV2.cairo").class_hash
+    %}
+
+    %{ expect_revert(error_type="ENTRY_POINT_NOT_FOUND_IN_CONTRACT") %}
+    IV2.test_v2_contract(contract_address=factory_address);
+
+    let (fee_to_setter_address_initial) = IFactory.get_fee_to_setter(contract_address=factory_address);
+
+    %{ stop_prank = start_prank(ids.deployer_address, target_contract_address=ids.factory_address) %}
+    IProxy.upgrade(
+        contract_address=factory_address, new_implementation=declared_factory_v2_class_hash
+    );
+    %{ stop_prank() %}
+
+    let (factory_implementation) = IProxy.get_implementation_hash(contract_address=factory_address);
+    assert factory_implementation = declared_factory_v2_class_hash;
+
+    let (factory_v2_success) = IV2.test_v2_contract(contract_address=factory_address);
```

Comment:
Ok, this is why we need that exact copy with the extra `test_v2_contract` function.
But could we not just check deployed addresses to make sure the "implementation contract" has been updated?